File: smartlist/src/main/java/com/smartlist/module/grading/GradingService.java
package com.smartlist.module.grading;

import com.smartlist.enums.AssessmentDecision;
import com.smartlist.enums.AssessmentStatus;
import com.smartlist.enums.CandidateAssessmentStatus;
import com.smartlist.enums.QuestionType;
import com.smartlist.model.*;
import com.smartlist.module.assessment.AssessmentRepository;
import com.smartlist.module.assessmentblock.AssessmentBlockRepository;
import com.smartlist.module.candidate.CandidateAssessmentRepository;
import com.smartlist.module.candidate.CandidateRepository;
import com.smartlist.module.candidate.CandidateResponseRepository;
import com.smartlist.module.candidate.interfaces.IQuestionStrategy;
import com.smartlist.module.grading.dto.request.QuestionCommentReqDTO;
import com.smartlist.module.grading.dto.request.QuestionScoreReqDTO;
import com.smartlist.module.grading.dto.response.AssessmentQuestionScoreRespDTO;
import com.smartlist.module.grading.dto.response.QuestionCommentRespDTO;
import com.smartlist.module.question.QuestionRepository;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;

@Service
public class GradingService {

    private final CandidateRepository candidateRepository;
    private final AssessmentRepository assessmentRepository;
    private final AssessmentBlockRepository assessmentBlockRepository;
    private final GradingRepository gradingRepository;
    private final QuestionRepository questionRepository;
    private final QuestionCommentRepository questionCommentRepository;
    private final CandidateResponseRepository candidateResponseRepository;
    private final CandidateAssessmentRepository candidateAssessmentRepository;
    private final ModelMapper modelMapper;

    public GradingService(final GradingRepository gradingRepository, final QuestionRepository questionRepository,
                          final QuestionCommentRepository questionCommentRepository, final CandidateResponseRepository candidateResponseRepository,
                          final CandidateAssessmentRepository candidateAssessmentRepository, final CandidateRepository candidateRepository,
                          final AssessmentRepository assessmentRepository, final AssessmentBlockRepository assessmentBlockRepository,
                          final ModelMapper modelMapper){
        this.gradingRepository = gradingRepository;
        this.questionRepository = questionRepository;
        this.questionCommentRepository = questionCommentRepository;
        this.candidateResponseRepository = candidateResponseRepository;
        this.candidateAssessmentRepository = candidateAssessmentRepository;
        this.candidateRepository = candidateRepository;
        this.assessmentRepository = assessmentRepository;
        this.assessmentBlockRepository = assessmentBlockRepository;
        this.modelMapper = modelMapper;
    }

    public List<AssessmentQuestionScoreRespDTO> getCandidateScoresByAssessment(final Integer candidateId, final Integer assessmentId) {
        Candidate candidate  = getCandidateByIdOrThrow(candidateId);
        Assessment assessment = getAssessmentByIdOrThrow(assessmentId);

        List<CandidateResponseScore> candidateResponseScores = this.gradingRepository.findByCandidateAndAssessment(candidate, assessment);
        List<AssessmentQuestionScoreRespDTO> assessmentQuestionScoreRespDTOs = new ArrayList<>();

        candidateResponseScores.forEach(candidateResponseScore -> assessmentQuestionScoreRespDTOs.add(this.modelMapper.map(candidateResponseScore, AssessmentQuestionScoreRespDTO.class)));
        return assessmentQuestionScoreRespDTOs;
    }

    public List<AssessmentQuestionScoreRespDTO> getCandidateScoresByAssessmentBlock(final Integer candidateId, final Integer assessmentBlockId) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        AssessmentBlock assessmentBlock = getAssessmentBlockByIdOrThrow(assessmentBlockId);

        List<CandidateResponseScore> candidateResponseScores = this.gradingRepository.findByCandidateAndAssessmentBlock(candidate, assessmentBlock);
        List<AssessmentQuestionScoreRespDTO> assessmentQuestionScoreRespDTOs = new ArrayList<>();

        candidateResponseScores.forEach(candidateResponseScore -> assessmentQuestionScoreRespDTOs.add(this.modelMapper.map(candidateResponseScore, AssessmentQuestionScoreRespDTO.class)));
        return assessmentQuestionScoreRespDTOs;
    }

    public List<Integer> getCandidateUnScoredQuestionsForAssessment(final Integer candidateId, final Integer assessmentId) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Assessment assessment = getAssessmentByIdOrThrow(assessmentId);

        List<Question> questions = this.questionRepository.findByAssessment(assessment);

        List<CandidateResponseScore> candidateResponseScores = this.gradingRepository.findByCandidateAndAssessment(candidate, assessment);

        ArrayList<Integer> notScoredQuestionIds  = new ArrayList<>();
        for (Question question : questions) {
            if (question.getType() != QuestionType.TEXT_RESPONSE
                    && question.getType() != QuestionType.VIDEO_RESPONSE
                    && question.getType() != QuestionType.FILE_RESPONSE) {
                continue;
            }
            int id = question.getId();
            Optional<CandidateResponseScore> optionalCandidateResponseScore
                    = candidateResponseScores.stream().filter(crs -> crs.getQuestion().getId() == id).findFirst();
            if (optionalCandidateResponseScore.isEmpty()) {
                notScoredQuestionIds.add(question.getId());
            }
        }
        return notScoredQuestionIds;
    }

    public List<QuestionCommentRespDTO> getCandidateQuestionComments(final Integer candidateId, final Integer questionId) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Question question = getQuestionByIdOrThrow(questionId);

        List<QuestionComment> questionComments = this.questionCommentRepository.findByCandidateAndQuestion(candidate, question);

        List<QuestionCommentRespDTO> questionCommentRespDTOS = new ArrayList<>();
        questionComments.forEach(questionComment -> questionCommentRespDTOS.add(this.modelMapper.map(questionComment, QuestionCommentRespDTO.class)));

        return questionCommentRespDTOS;
    }

    public void saveCandidateQuestionComment(final Integer candidateId, final Integer questionId, final QuestionCommentReqDTO questionCommentReqDto) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Question question = getQuestionByIdOrThrow(questionId);

        QuestionComment questionComment = this.questionCommentRepository.findFirstByCandidateAndQuestionAndUsername(candidate, question, questionCommentReqDto.getUsername());
        if (questionComment != null) {
            questionComment.setComment(questionCommentReqDto.getComment());
        } else {
            questionComment = new QuestionComment();
            questionComment.setCandidate(candidate);
            questionComment.setQuestion(question);
            questionComment.setUsername(questionCommentReqDto.getUsername());
            questionComment.setComment(questionCommentReqDto.getComment());
        }

        this.questionCommentRepository.save(questionComment);
    }

    @Transactional
    public void saveCandidateResponseScore(final Integer candidateId, final QuestionScoreReqDTO questionScoreReqDto) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Assessment assessment = getAssessmentByIdOrThrow(questionScoreReqDto.getAssessmentId());

        if (assessment.getStatus() != AssessmentStatus.ACTIVE) {
            throw GradingResponseCodes.ASSESSMENT_ACTION_DENIED;
        }

        Optional<Question> optionalQuestion = this.questionRepository.findById(questionScoreReqDto.getQuestionId());
        Question question = optionalQuestion.orElseThrow(() -> GradingResponseCodes.INVALID_QUESTION_ID);

        if (!Objects.equals(question.getAssessmentBlock().getAssessment().getId(), assessment.getId())) {
            throw GradingResponseCodes.INVALID_ASSESSMENT_ID;
        }

        CandidateResponseScore candidateResponseScore = new CandidateResponseScore(candidate, question);
        candidateResponseScore.setScore(questionScoreReqDto.getScore());

        this.gradingRepository.removeByCandidateAndQuestion(candidate, question);
        this.gradingRepository.save(candidateResponseScore);
    }

    public void markCandidateAssessmentGradingComplete(final Integer candidateId, final Integer assessmentId) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Assessment assessment = getAssessmentByIdOrThrow(assessmentId);

        if (assessment.getStatus() != AssessmentStatus.ACTIVE) {
            throw GradingResponseCodes.ASSESSMENT_ACTION_DENIED;
        }

        Optional<CandidateAssessment> optionalCandidateAssessment = this.candidateAssessmentRepository.findByCandidateAndAssessment(candidate, assessment);
        CandidateAssessment candidateAssessment = optionalCandidateAssessment.orElseThrow(() -> GradingResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS);

        if (candidateAssessment.getStatus() != CandidateAssessmentStatus.GRADING_PENDING) {
            throw GradingResponseCodes.CANDIDATE_ASSESSMENT_ACTION_DENIED;
        }

        int countOfScoredResponses = this.gradingRepository.countByCandidateAndAssessment(candidate, assessment);
        int countOfResponses = this.candidateResponseRepository.countByCandidateAssessment(candidateAssessment);

        if (countOfScoredResponses != countOfResponses) {
            throw GradingResponseCodes.CANDIDATE_ASSESSMENT_SCORING_PENDING;
        }

        candidateAssessment.setStatus(CandidateAssessmentStatus.GRADING_COMPLETED);
        candidateAssessment.setEndDate(LocalDateTime.now());
        this.candidateAssessmentRepository.save(candidateAssessment);
    }

    public AssessmentDecision getCandidateAssessmentDecision(final Integer candidateId, final Integer assessmentId) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Assessment assessment = getAssessmentByIdOrThrow(assessmentId);

        Optional<CandidateAssessment> optionalCandidateAssessment = this.candidateAssessmentRepository.findByCandidateAndAssessment(candidate, assessment);

        if (optionalCandidateAssessment.isEmpty()) {
            throw GradingResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS;
        }

        return optionalCandidateAssessment.get().getAssessmentDecision();
    }

    public void setCandidateAssessmentDecision(final Integer candidateId, final Integer assessmentId, final String decision) {
        Candidate candidate = getCandidateByIdOrThrow(candidateId);
        Assessment assessment = getAssessmentByIdOrThrow(assessmentId);

        if (assessment.getStatus() != AssessmentStatus.ACTIVE) {
            throw GradingResponseCodes.ASSESSMENT_ACTION_DENIED;
        }

        AssessmentDecision assessmentDecision;
        try {
            assessmentDecision = AssessmentDecision.valueOf(decision);
        }catch (IllegalArgumentException e){
            throw GradingResponseCodes.INVALID_ASSESSMENT_DECISION;
        }

        Optional<CandidateAssessment> optionalCandidateAssessment = this.candidateAssessmentRepository.findByCandidateAndAssessment(candidate, assessment);
        CandidateAssessment candidateAssessment = optionalCandidateAssessment.orElseThrow(() -> GradingResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS);

        if (candidateAssessment.getStatus() != CandidateAssessmentStatus.GRADING_COMPLETED) {
            throw GradingResponseCodes.CANDIDATE_ASSESSMENT_ACTION_DENIED;
        }

        candidateAssessment.setAssessmentDecision(assessmentDecision);
        this.candidateAssessmentRepository.save(candidateAssessment);
    }

    public void autoScoreAndSaveCandidateResponseScore(final Candidate candidate, final Question question, final IQuestionStrategy questionStrategy) {
        CandidateResponseScore candidateResponseScore = new CandidateResponseScore();
        candidateResponseScore.setCandidate(candidate);
        candidateResponseScore.setQuestion(question);
        candidateResponseScore.setAssessmentBlock(question.getAssessmentBlock());
        candidateResponseScore.setAssessment(question.getAssessmentBlock().getAssessment());
        candidateResponseScore.setScore(questionStrategy.score());

        this.gradingRepository.removeByCandidateAndQuestion(candidate, question);
        this.gradingRepository.save(candidateResponseScore);
    }

    private Candidate getCandidateByIdOrThrow(final Integer candidateId) {
        final Optional<Candidate> optionalCandidate = this.candidateRepository.findById(candidateId);
        return optionalCandidate.orElseThrow(() -> GradingResponseCodes.INVALID_CANDIDATE_ID);
    }

    private Question getQuestionByIdOrThrow(final Integer questionId) {
        final Optional<Question> optionalQuestion = this.questionRepository.findById(questionId);
        return optionalQuestion.orElseThrow(() -> GradingResponseCodes.INVALID_QUESTION_ID);
    }


    private AssessmentBlock getAssessmentBlockByIdOrThrow(final Integer assessmentBlockId) {
        final Optional<AssessmentBlock> optionalAssessmentBlock = this.assessmentBlockRepository.findById(assessmentBlockId);
        return optionalAssessmentBlock.orElseThrow(() -> GradingResponseCodes.INVALID_ASSESSMENT_BLOCK_ID);
    }

    private Assessment getAssessmentByIdOrThrow(final Integer id) {
        Optional<Assessment> optionalAssessment = this.assessmentRepository.findById(id);
        return optionalAssessment.orElseThrow (() -> GradingResponseCodes.INVALID_ASSESSMENT_ID);
    }
}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/grading/grading.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { AssessmentBlock } from 'src/entities/assessment-block.entity';
import { Assessment } from 'src/entities/assessment.entity';
import { CandidateAssessment } from 'src/entities/candidate-assessment.entity';
import { CandidateResponseScore } from 'src/entities/candidate-response-score.entity';
import { Candidate } from 'src/entities/candidate.entity';
import { QuestionComment } from 'src/entities/question-comment.entity';
import { Question } from 'src/entities/question.entity';
import { AssessmentDecision } from 'src/enums/assessment.decision';
import { AssessmentStatus } from 'src/enums/assessment.status';
import { CandidateAssessmentStatus } from 'src/enums/candidate.assessment.status';
import { QuestionType } from 'src/enums/question.type';

import { AssessmentBlockRepository } from '../assessment-block/assessment-block.repository';
import { AssessmentRepository } from '../assessment/assessment.repository';
import { CandidateAssessmentRepository } from '../candidate/candidate.assessment.repository';
import { CandidateRepository } from '../candidate/candidate.repository';
import { CandidateResponseRepository } from '../candidate/candidate.response.repository';
import { IQuestionStrategy } from '../candidate/interfaces/interface-question-strategy';
import { QuestionRepository } from '../question/question.repository';
import { QuestionCommentReqDto } from './dto/request/question-comment.req.dto';
import { QuestionScoreReqDto } from './dto/request/question-score.req.dto';
import { AssessmentQuestionScoreRespDto } from './dto/response/assessment-question-score.resp.dto';
import { QuestionCommentRespDto } from './dto/response/question-comment.resp.dto';
import { GradingRepository } from './grading.repository';
import { GradingResponseCodes } from './grading.response.codes';
import { QuestionCommentRepository } from './question-comment.repository';

@Injectable()
export class GradingService {
  constructor(
    @InjectRepository(GradingRepository)
    private gradingRepository: GradingRepository,
    @InjectRepository(CandidateRepository)
    private candidateRepository: CandidateRepository,
    @InjectRepository(AssessmentRepository)
    private assessmentRepository: AssessmentRepository,
    @InjectRepository(AssessmentBlockRepository)
    private assessmentBlockRepository: AssessmentBlockRepository,
    @InjectRepository(QuestionRepository)
    private questionRepository: QuestionRepository,
    @InjectRepository(CandidateAssessmentRepository)
    private candidateAssessmentRepository: CandidateAssessmentRepository,
    @InjectRepository(QuestionCommentRepository)
    private questionCommentRepository: QuestionCommentRepository,
    @InjectRepository(CandidateResponseRepository)
    private candidateResponseRepository: CandidateResponseRepository,
  ) {}

  async getCandidateScoresByAssessment(
    candidateId: number,
    assessmentId: number,
  ): Promise<AssessmentQuestionScoreRespDto[]> {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessment: Assessment = await this.getAssessmentByIdOrThrow(
      assessmentId,
    );

    const candidateResponseScores: CandidateResponseScore[] =
      await this.gradingRepository.getForCandidateByAssessment(
        candidate,
        assessment,
      );
    const assessmentQuestionScoreRespDtos: AssessmentQuestionScoreRespDto[] =
      [];

    candidateResponseScores.forEach((candidateResponseScore) => {
      assessmentQuestionScoreRespDtos.push(
        new AssessmentQuestionScoreRespDto(candidateResponseScore),
      );
    });
    return assessmentQuestionScoreRespDtos;
  }

  async getCandidateScoresByAssessmentBlock(
    candidateId: number,
    assessmentBlockId: number,
  ): Promise<AssessmentQuestionScoreRespDto[]> {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessmentBlock: AssessmentBlock =
      await this.getAssessmentBlockByIdOrThrow(assessmentBlockId);

    const candidateResponseScores: CandidateResponseScore[] =
      await this.gradingRepository.getForCandidateByAssessmentBlock(
        candidate,
        assessmentBlock,
      );
    const assessmentQuestionScoreRespDtos: AssessmentQuestionScoreRespDto[] =
      [];

    candidateResponseScores.forEach((candidateResponseScore) => {
      assessmentQuestionScoreRespDtos.push(
        new AssessmentQuestionScoreRespDto(candidateResponseScore),
      );
    });
    return assessmentQuestionScoreRespDtos;
  }

  async getCandidateUnScoredQuestionsForAssessment(
    candidateId: number,
    assessmentId: number,
  ): Promise<number[]> {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessment: Assessment = await this.getAssessmentByIdOrThrow(
      assessmentId,
    );

    const questions: Question[] =
      await this.questionRepository.findAllByAssessment(assessment);

    const candidateResponseScores: CandidateResponseScore[] =
      await this.gradingRepository.getForCandidateByAssessment(
        candidate,
        assessment,
      );

    const unscoredQuestionIds: number[] = [];
    for (let i = 0; i < questions.length; i++) {
      if (
        questions[i].type != QuestionType.TEXT_RESPONSE &&
        questions[i].type != QuestionType.VIDEO_RESPONSE &&
        questions[i].type != QuestionType.FILE_RESPONSE
      ) {
        continue;
      }
      const candidateResponseScore: CandidateResponseScore =
        candidateResponseScores.find(
          (crs) => crs.question.id == questions[i].id,
        );
      if (candidateResponseScore == null) {
        unscoredQuestionIds.push(questions[i].id);
      }
    }
    return unscoredQuestionIds;
  }

  async getCandidateQuestionComments(
    candidateId: number,
    questionId: number,
  ): Promise<QuestionCommentRespDto[]> {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const question: Question = await this.getQuestionByIdOrThrow(questionId);

    const questionComments: QuestionComment[] =
      await this.questionCommentRepository.findAllForCandidateQuestion(
        candidate,
        question,
      );

    const questionCommentRespDtos: QuestionCommentRespDto[] = [];
    questionComments.forEach((questionComment) => {
      questionCommentRespDtos.push(new QuestionCommentRespDto(questionComment));
    });

    return questionCommentRespDtos;
  }

  async saveCandidateQuestionComment(
    candidateId: number,
    questionId: number,
    questionCommentReqDto: QuestionCommentReqDto,
  ) {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const question: Question = await this.getQuestionByIdOrThrow(questionId);

    let questionComment: QuestionComment =
      await this.questionCommentRepository.getForCandidateQuestionAndUsername(
        candidate,
        question,
        questionCommentReqDto.username,
      );
    if (questionComment != null) {
      questionComment.comment = questionCommentReqDto.comment;
    } else {
      questionComment = new QuestionComment();
      questionComment.candidate = candidate;
      questionComment.question = question;
      questionComment.username = questionCommentReqDto.username;
      questionComment.comment = questionCommentReqDto.comment;
    }

    this.questionCommentRepository.save(questionComment);
  }

  async saveCandidateResponseScore(
    candidateId: number,
    questionScoreReqDto: QuestionScoreReqDto,
  ) {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessment: Assessment = await this.getAssessmentByIdOrThrow(
      questionScoreReqDto.assessment_id,
    );

    if (assessment.status != AssessmentStatus.ACTIVE) {
      throw GradingResponseCodes.ASSESSMENT_ACTION_DENIED;
    }

    const question: Question =
      await this.questionRepository.findByIdWithRelations(
        questionScoreReqDto.question_id,
      );
    if (question == null) {
      throw GradingResponseCodes.INVALID_QUESTION_ID;
    }

    if (question.assessmentBlock.assessment.id != assessment.id) {
      throw GradingResponseCodes.INVALID_ASSESSMENT_ID;
    }

    const candidateResponseScore: CandidateResponseScore =
      new CandidateResponseScore();
    candidateResponseScore.candidate = candidate;
    candidateResponseScore.question = question;
    candidateResponseScore.assessmentBlock = question.assessmentBlock;
    candidateResponseScore.assessment = question.assessmentBlock.assessment;
    candidateResponseScore.score = questionScoreReqDto.score;

    await this.gradingRepository.delete({
      candidate: candidate,
      question: question,
    });
    await this.gradingRepository.save(candidateResponseScore);
  }

  async autoScoreAndSaveCandidateResponseScore(
    candidate: Candidate,
    question: Question,
    questionStrategy: IQuestionStrategy,
  ) {
    const candidateResponseScore: CandidateResponseScore =
      new CandidateResponseScore();
    candidateResponseScore.candidate = candidate;
    candidateResponseScore.question = question;
    candidateResponseScore.assessmentBlock = question.assessmentBlock;
    candidateResponseScore.assessment = question.assessmentBlock.assessment;
    candidateResponseScore.score = questionStrategy.score();

    await this.gradingRepository.delete({
      candidate: candidate,
      question: question,
    });

    await this.gradingRepository.save(candidateResponseScore);
  }

  async markCandidateAssessmentGradingComplete(
    candidateId: number,
    assessmentId: number,
  ) {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessment: Assessment = await this.getAssessmentByIdOrThrow(
      assessmentId,
    );

    if (assessment.status != AssessmentStatus.ACTIVE) {
      throw GradingResponseCodes.ASSESSMENT_ACTION_DENIED;
    }

    const candidateAssessment: CandidateAssessment =
      await this.candidateAssessmentRepository.getForCandidateAndAssessment(
        candidate,
        assessment,
      );

    if (!candidateAssessment) {
      throw GradingResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS;
    }
    if (
      candidateAssessment.status != CandidateAssessmentStatus.GRADING_PENDING
    ) {
      throw GradingResponseCodes.CANDIDATE_ASSESSMENT_ACTION_DENIED;
    }

    const countOfScoredResponses: number =
      await this.gradingRepository.getCountForCandidateByAssessment(
        candidate,
        assessment,
      );

    const countOfResponses: number =
      await this.candidateResponseRepository.findCountByCandidateAssessment(
        candidateAssessment,
      );

    if (countOfScoredResponses != countOfResponses) {
      throw GradingResponseCodes.CANDIDATE_ASSESSMENT_SCORING_PENDING;
    }

    candidateAssessment.status = CandidateAssessmentStatus.GRADING_COMPLETED;
    candidateAssessment.endDate = new Date();
    await this.candidateAssessmentRepository.save(candidateAssessment);
  }

  async getCandidateAssessmentDecision(
    candidateId: number,
    assessmentId: number,
  ): Promise<AssessmentDecision> {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessment: Assessment = await this.getAssessmentByIdOrThrow(
      assessmentId,
    );

    const candidateAssessment: CandidateAssessment =
      await this.candidateAssessmentRepository.getForCandidateAndAssessment(
        candidate,
        assessment,
      );

    if (!candidateAssessment) {
      throw GradingResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS;
    }

    return candidateAssessment.assessmentDecision;
  }

  async setCandidateAssessmentDecision(
    candidateId: number,
    assessmentId: number,
    decision: string,
  ) {
    const candidate: Candidate = await this.getCandidateByIdOrThrow(
      candidateId,
    );

    const assessment: Assessment = await this.getAssessmentByIdOrThrow(
      assessmentId,
    );

    if (assessment.status != AssessmentStatus.ACTIVE) {
      throw GradingResponseCodes.ASSESSMENT_ACTION_DENIED;
    }

    const assessmentDecision: AssessmentDecision =
      AssessmentDecision[decision as keyof typeof AssessmentDecision];

    if (assessmentDecision == null) {
      throw GradingResponseCodes.INVALID_ASSESSMENT_DECISION;
    }

    const candidateAssessment: CandidateAssessment =
      await this.candidateAssessmentRepository.getForCandidateAndAssessment(
        candidate,
        assessment,
      );

    if (candidateAssessment == null) {
      throw GradingResponseCodes.CANDIDATE_ASSESSMENT_DOES_NOT_EXISTS;
    }

    if (
      candidateAssessment.status != CandidateAssessmentStatus.GRADING_COMPLETED
    ) {
      throw GradingResponseCodes.CANDIDATE_ASSESSMENT_ACTION_DENIED;
    }

    candidateAssessment.assessmentDecision = assessmentDecision;

    await this.candidateAssessmentRepository.save(candidateAssessment);
  }

  private async getCandidateByIdOrThrow(id: number): Promise<Candidate> {
    const candidate: Candidate = await this.candidateRepository.findById(id);

    if (candidate == null) {
      throw GradingResponseCodes.INVALID_CANDIDATE_ID;
    }
    return candidate;
  }

  private async getAssessmentByIdOrThrow(id: number): Promise<Assessment> {
    const assessment: Assessment = await this.assessmentRepository.findById(id);
    if (assessment == null) {
      throw GradingResponseCodes.INVALID_ASSESSMENT_ID;
    }
    return assessment;
  }

  private async getAssessmentBlockByIdOrThrow(
    id: number,
  ): Promise<AssessmentBlock> {
    const assessmentBlock: AssessmentBlock =
      await this.assessmentBlockRepository.findById(id);
    if (assessmentBlock == null) {
      throw GradingResponseCodes.INVALID_ASSESSMENT_BLOCK_ID;
    }
    return assessmentBlock;
  }

  private async getQuestionByIdOrThrow(id: number): Promise<Question> {
    const question: Question = await this.questionRepository.findById(id);
    if (question == null) {
      throw GradingResponseCodes.INVALID_QUESTION_ID;
    }
    return question;
  }
}
